### ISSUE TITLE
Extraheader with Gadget

### ISSUE THREAD

**Issue by loikki** [yt-project/yt#1846]
### Bug report

**Bug summary**

When reading a gadget file with an extra header (e.g. https://obswww.unige.ch/~lhausamm/swift/IC/AgoraDisk/Gear/snapshot_0000), you do not take it into account here https://github.com/yt-project/yt/blob/master/yt/frontends/gadget/data_structures.py#L396

**Code for reproduction**

I am using the AGORA analysis script (https://bitbucket.org/mornkr/agora-analysis-script/src/8f12b112e19614d66274ece5a5bf21af23a76778/script-Paper4-disk-new.py?at=default&fileviewer=file-view-default) where they define the extra header here https://bitbucket.org/mornkr/agora-analysis-script/src/8f12b112e19614d66274ece5a5bf21af23a76778/script-Paper4-disk-new.py?at=default&fileviewer=file-view-default#script-Paper4-disk-new.py-150 and the code fails when adding a field here https://bitbucket.org/mornkr/agora-analysis-script/src/8f12b112e19614d66274ece5a5bf21af23a76778/script-Paper4-disk-new.py?at=default&fileviewer=file-view-default#script-Paper4-disk-new.py-501

**Actual outcome**

```
Traceback (most recent call last):
  File "./plot_solution.py", line 519, in <module>
    pf.add_field(("gas", "density_squared"), function=_density_squared, units="g**2/cm**6")
  File "/usr/local/lib/python2.7/dist-packages/yt/data_objects/static_output.py", line 1191, in add_field
    self.index
  File "/usr/local/lib/python2.7/dist-packages/yt/data_objects/static_output.py", line 504, in index
    self, dataset_type=self.dataset_type)
  File "/usr/local/lib/python2.7/dist-packages/yt/geometry/particle_geometry_handler.py", line 39, in __init__
    super(ParticleIndex, self).__init__(ds, dataset_type)
  File "/usr/local/lib/python2.7/dist-packages/yt/geometry/geometry_handler.py", line 50, in __init__
    self._setup_geometry()
  File "/usr/local/lib/python2.7/dist-packages/yt/geometry/particle_geometry_handler.py", line 50, in _setup_geometry
    self._initialize_particle_handler()
  File "/usr/local/lib/python2.7/dist-packages/yt/geometry/particle_geometry_handler.py", line 104, in _initialize_particle_handler
    self._initialize_indices()
  File "/usr/local/lib/python2.7/dist-packages/yt/geometry/particle_geometry_handler.py", line 134, in _initialize_indices
    self.io._initialize_index(data_file, self.regions)
  File "/usr/local/lib/python2.7/dist-packages/yt/frontends/gadget/io.py", line 351, in _initialize_index
    data_file, count, 0, regions, DLE, DRE)
  File "/usr/local/lib/python2.7/dist-packages/yt/frontends/gadget/io.py", line 334, in _get_morton_from_position
    pp = np.fromfile(f, dtype=self._endian + self._float_type,
TypeError: cannot concatenate 'str' and 'int' objects

```

**Expected outcome**

**Version Information**

  * Operating System: Linux Ubuntu 16.04
  * Python Version: 2.7.12
  * yt version: 3.4.1 (installed through pip)
  * Other Libraries (if applicable):

**Comment by ngoldbaum** [comment]
Thanks for the report and steps to reproduce. I will try to take a look at this soon.

**Comment by qobilidop** [comment]
Hi @loikki, I ran into the same issue with another dataset. Basically, I can trigger the issue by

```python
from yt.frontends.gadget.api import GadgetDataset
ds = GadgetDataset('path/to/snapshot')
ds.field_list
```

But my dataset is too large (> 30G) to share. I'm wondering if you could share your dataset again (the first link is broken)? I'd like to take a deeper look and hopefully come up with a fix.

**Comment by MatthewTurk** [comment]
We do have a number of twiddles for specifying additional fields and whatnot (`gadget_header_spec` https://github.com/yt-project/yt/blob/master/yt/frontends/gadget/definitions.py#L17 ), but it looks like this falls into a corner case.

**Comment by qobilidop** [comment]
@MatthewTurk I guess the fix would then be making `gadget_header_specs` configurable in `GadgetDataset`'s constructor and check the code to use a calculated header size instead of 256?

**Comment by qobilidop** [comment]
@MatthewTurk I guess the fix would be to check the code and use a calculated header size instead of 256, for example in this [line](https://github.com/yt-project/yt/blob/15c0691d9de2abdf320eaaccab5d2b8ac3af4aef/yt/frontends/gadget/data_structures.py#L396) as @loikki pointed out.

**Comment by ngoldbaum** [comment]
Sure, if real-world datasets don't have that exact amount of padding, we should inspect the data and calculate it.

FWIW I'm pretty sure that constant comes from the documentation for the the Gadget-2 output format in the user guide, I think it was fair of yt to expect real-world datasets to conform to the format documented there although if that's not the case we should definitely fix it :)

Presumably there's a way to do this either by looking very low-level at the bytes in the header or through creative use of the `struct` module.

**Comment by qobilidop** [comment]
@ngoldbaum Yeah, 256 is the Gadget standard. But since [`header_spec` is already configurable](https://github.com/yt-project/yt/blob/15c0691d9de2abdf320eaaccab5d2b8ac3af4aef/yt/frontends/gadget/data_structures.py#L107), to be self-consistent, I think we should use a caculated header size.

**Comment by loikki** [comment]
Sorry for the late answer.

I have reduced in size my snapshot (less than 1Mo) and you can download it here https://obswww.unige.ch/~lhausamm/git/test

I am using the following script to read it:
```python
from yt.frontends.gadget.definitions import gadget_header_specs
from yt.frontends.gadget.definitions import gadget_field_specs
from yt.mods import *

filename = "test"

gadget_header_specs["chemistry"] = (('h1',  4, 'c'), ('h2',  4, 'c'),
                                    ('empty', 256, 'c'),)
gear_ptype_specs = ("Gas", "Stars", "Halo", "Disk", "Bulge", "Bndry")

gadget_default_unit_base = {
    'UnitLength_in_cm': 3.08568e+21,
    'UnitMass_in_g': 1.989e+43,
    'UnitVelocity_in_cm_per_s': 100000
}

pf = GadgetDataset(filename, unit_base=gadget_default_unit_base,
                   bounding_box=[[-1000.0, 1000.0], [-1000.0, 1000.0],
                                 [-1000.0, 1000.0]],
                   header_spec="default+chemistry",
                   ptype_spec=gear_ptype_specs)
```

**Comment by ngoldbaum** [comment]
Hmm maybe I merged Bili's PR prematurely, I didn't see this comment.

@qobilidop when I run @loikki's test script using the current master branch, I get:

```
Traceback (most recent call last):
  File "testg.py", line 21, in <module>
    ptype_spec=gear_ptype_specs)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/frontends/gadget/data_structures.py", line 171, in __init__
    kernel_name=kernel_name)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/frontends/sph/data_structures.py", line 36, in __init__
    n_ref=n_ref, over_refine_factor=over_refine_factor)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/static_output.py", line 1502, in __init__
    units_override=units_override, unit_system=unit_system)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/static_output.py", line 309, in __init__
    self._parse_parameter_file()
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/frontends/gadget/data_structures.py", line 208, in _parse_parameter_file
    hvals = self._get_hvals()
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/frontends/gadget/data_structures.py", line 196, in _get_hvals
    gformat = _get_gadget_format(self.parameter_filename, self._header_size)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/frontends/gadget/data_structures.py", line 75, in _get_gadget_format
    raise RuntimeError("Incorrect Gadget format %s!" % str(rhead))
RuntimeError: Incorrect Gadget format 256!
```

Is that what you'd expect? Is there some modification to the test script that needs to happen before the data loads?

**Comment by ngoldbaum** [comment]
Sorry, to clarify, is the conclusion you came to that the file is corrupt somehow and Loic should fix it?

Given the real-world existence of this data, is there something we can do to detect this case, maybe print a warning, and load the data anyway? Or am I misunderstanding?

**Comment by qobilidop** [comment]
Yes, my conclusion is that there should be a single header block in the snapshot file. As long as that's the case, yt shall be able to load correctly.

But before that, there are several things I'm uncertain about, especially the interpretation of the header spec. Does it really mean to ignore the block size bytes between the two blocks? Maybe @loikki could clarify.

We could make this snapshot loadable by ignoring the block size bytes and rely on header spec solely. I just worry that without the checking of block size, there will be other issues.

**Comment by loikki** [comment]
We do have two headers. The decision was motivated to make the header more compatible with other softwares (in this way, we just need to skip the second header and not modify the first reading).

I am fine with any of your decision, I can easily trick yt to work on my old snapshot.
I guess that we can improve the error message if you wish to allow only a single header.

**Comment by qobilidop** [comment]
@loikki Thanks for the information!

So it comes down to how we want to support multiple headers. I have two proposals:

1. The cleaner way: For multiple headers, we pass a list to `header_spec`. `header_spec='default+chemistry'` means a single long header, while `header_spec=['default', 'chemistry']` means 2 headers.
2. The compatible way: We make it optional to turn off the header size check, and define some extra fields in `header_spec` (like the `h1`, `h2` in the example) to jump between headers. This would make the example script work out of the box. But it's a little bit hacky in my opinion.

Which one do people prefer? Are there other proposals? I'll make a new PR to support multiple headers once the design descision has been made. @ngoldbaum @MatthewTurk

**Comment by ngoldbaum** [comment]
Given the status quo is that things don't work, I don't think we're bound by backward compatibility guarantees (unless this used to work?). In that case I'd go with what you think is the cleaner option, #1.

**Comment by loikki** [comment]
I have no idea if it was working or not before, but I approve the first possibility, it is the cleanest way to do it.

Thank you very much for your help and time!
